### ISSUE TITLE
"Speed" field of /core/stats api response gets inaccurate over time

### ISSUE THREAD

**Issue by JoshuaDietz** [rclone/rclone#4569]
#### What is the problem you are having with rclone?
The field "speed" of the API endpoint  /core/stats seems to drift over time. I'm uploading since 8 days straight now and currently the value is at about 400 MBPS (406479450.72536063). This is absolutely impossible because my upload is 80 times slower. When starting rclone the problem does not occur so this looks like the sum of a few million precision errors.

#### What is your rclone version (output from `rclone version`)
v1.52.3-309-g47d08ac1-beta. I can't use the newest one since I have to let it run for a few days for the error to occur :-)

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Raspbian OS 32bit

####  Which cloud storage system are you using? (eg Google Drive)
Jottacloud

**Comment by ncw** [comment]
Could you capture the output of the stats RC call and post two separated by some time so I can see the changes between them?

I wonder if something has wrapped since you are on a 32 bit OS.

**Comment by JoshuaDietz** [comment]
Sure @ncw 
This is after Running for about 8 days:

```
{
	"bytes": 412581966675,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 757870.944624575,
	"errors": 110,
	"fatalError": false,
	"lastError": "HTTP error 500 (500 Server Error) returned body: \"{\\\"code\\\":500,\\\"message\\\":\\\"CircuitBreaker 'dfs-044' is OPEN and does not permit further calls\\\",\\\"cause\\\":\\\"\\\",\\\"error_id\\\":\\\"CallNotPermittedException\\\",\\\"x-id\\\":\\\"185627190016\\\"}\"",
	"renames": 0,
	"retryError": true,
	"speed": 272946257.4279051,
	"transferTime": 1511.587050473,
	"transferring": [
		{
			"bytes": 1568768,
			"eta": 37,
			"group": "global_stats",
			"name": "backup1/my-devices/data/fe/fe3af393f90d7883abbe4bdbe9149df856cce515f49eccd93f1591d39b8f3f8b",
			"percentage": 30,
			"size": 5111846,
			"speed": 96157.90440109702,
			"speedAvg": 93940.25711127192
		},
		{
			"bytes": 1241088,
			"eta": 39,
			"group": "global_stats",
			"name": "backup1/my-devices/data/fe/fe3b55781912186d96d02e9d8145ef614eacef3628cfb2100982f19fa3f5d260",
			"percentage": 25,
			"size": 4782973,
			"speed": 94544.99778206991,
			"speedAvg": 90428.65243418742
		},
		{
			"bytes": 946176,
			"eta": 41,
			"group": "global_stats",
			"name": "backup1/my-devices/data/fe/fe3c7446bb5f0f9b744b870eddf37bbef21d4b31b232f53bfcc43a32b7227c85",
			"percentage": 20,
			"size": 4531731,
			"speed": 89395.95961418704,
			"speedAvg": 86016.35817710658
		},
		{
			"bytes": 684032,
			"eta": 62,
			"group": "global_stats",
			"name": "backup1/my-devices/data/fe/fe3d5cf2b1234fc38082e3034b985f0330839ea5767459682429035f0af86d31",
			"percentage": 11,
			"size": 5743321,
			"speed": 87212.39848478066,
			"speedAvg": 81407.65397464015
		}
	],
	"transfers": 81199
}

```

Then I restarted rclone and this response is after about 10 minutes of running again:

```
{
	"bytes": 228768767,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 638.134633702,
	"errors": 0,
	"fatalError": false,
	"renames": 0,
	"retryError": false,
	"speed": 362024.0405328464,
	"transferTime": 631.916098763,
	"transferring": [
		{
			"bytes": 2551808,
			"eta": 17,
			"group": "global_stats",
			"name": "backup1/my-devices/data/00/004d5cda3dabd3643381938869b50888a2d9796b3fa46e0ebbc7f654ffe5618a",
			"percentage": 60,
			"size": 4220391,
			"speed": 94100.83684081654,
			"speedAvg": 93105.75680740265
		},
		{
			"bytes": 2322432,
			"eta": 46,
			"group": "global_stats",
			"name": "backup1/my-devices/data/00/004da59d81d1c9795fc2e28ed6f106e66af33a40fe70044630ef982df320e64e",
			"percentage": 35,
			"size": 6498188,
			"speed": 91558.23235520121,
			"speedAvg": 90172.46263313154
		},
		{
			"bytes": 2158592,
			"eta": 23,
			"group": "global_stats",
			"name": "backup1/my-devices/data/00/004f88de98e3bd79c2bc7f27773bfcfbcb5bbf772aa611248725345e1bdbb909",
			"percentage": 51,
			"size": 4206337,
			"speed": 89292.0074692519,
			"speedAvg": 87408.0151479712
		},
		{
			"bytes": 2158592,
			"eta": 25,
			"group": "global_stats",
			"name": "backup1/my-devices/data/00/004fdff15052e18befd4bad3b7cbd6362dd7334d03f39ff4b0c51ca610e78b77",
			"percentage": 48,
			"size": 4413309,
			"speed": 89480.49668806602,
			"speedAvg": 87537.04999904906
		}
	],
	"transfers": 43
}
```




I did it in this order because else I'd have to wait for a few days to respond. But I can post the values of the current run again after some time if you want to have the values of the same run.


Please note that I'm using a timetable for the bandwith limits. So the second api response (which I did about 10 minutes after the restart of rclone) was not influenced by all the timeslots (because it was not running for a whole day) and the first one was (because it was running for 8 days). But even though my timetable has an unlimited-speed-slot for a few hours a day it is absolutely impossible that the speed is that high since the maximum upload of my internet connection is less than 10 mbit/s and since the first response was captured in a timeslot with a 350K speed limit.

And yes I can imagine too that it has to do something with the 32bit OS. But probably this only amplifies the problem.

**Comment by ncw** [comment]
Something definitely has gone a bit crazy there.

In the first stats the transfers add up to about 400k/s. The bytes / 8 days comes out to about 600k/s which doesn't seem unreasonable.

What does seem unreasonable is 

>	"elapsedTime": 757870.944624575,
> 	"transferTime": 1511.587050473

This is saying that you've been transferring data for about 25mins out of 8.7 days.

Are you transferring data continuously?

 In the second they look sensible.

> 	"elapsedTime": 638.134633702,
> 	"transferTime": 631.916098763,

Could you log those two values at regular intervals (say every hour) and see how they diverge? I guess they might either diverge slowly or jump when something happens.

**Comment by JoshuaDietz** [comment]
Okay, so here is the result after about 1,5h of running (still the same run as my second stats result above and still within my 350K speed limit timeslot)

```
{
	"bytes": 1872030707,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 5223.129277676,
	"errors": 0,
	"fatalError": false,
	"renames": 0,
	"retryError": false,
	"speed": 1289826.2391248,
	"transferTime": 1451.382347904,
	"transferring": [
		{
			"bytes": 5632000,
			"eta": 8,
			"group": "global_stats",
			"name": "backup1/my-devices/data/02/027000ec2ed7c57d62c6031b526891c5e92d03007c566387d2af147e55ea3966",
			"percentage": 88,
			"size": 6380723,
			"speed": 91280.37658432747,
			"speedAvg": 89773.26416227271
		},
		{
			"bytes": 4550656,
			"eta": 5,
			"group": "global_stats",
			"name": "backup1/my-devices/data/02/027137eef45f737bf71f69618b2a770c03243c62d4d4ff48ffa15409c8c9c91e",
			"percentage": 90,
			"size": 5012873,
			"speed": 90268.03407584176,
			"speedAvg": 89667.54602713198
		},
		{
			"bytes": 3960832,
			"eta": 18,
			"group": "global_stats",
			"name": "backup1/my-devices/data/02/02736fd37f075fec31fe942a87dd6810e788b10c97348312c3824a65b60a2662",
			"percentage": 69,
			"size": 5668879,
			"speed": 89906.4471207109,
			"speedAvg": 89993.93861722822
		},
		{
			"bytes": 2224128,
			"eta": 22,
			"group": "global_stats",
			"name": "backup1/my-devices/data/02/027548fdfc306fb71b72734a25d532045ca04ef68c13462003aff8cac9e73741",
			"percentage": 53,
			"size": 4194892,
			"speed": 88573.761112864,
			"speedAvg": 86350.0903363357
		}
	],
	"transfers": 373
}
```

I have zero experience in go and I'm not really familar with the codebase of the project. I looked around the code a bit anyway and I think that the cause of the problem is the function totalDuration [here](https://github.com/rclone/rclone/blob/f5439ddc54ae63f55ce9f23a743c18fc9d7dbd4a/fs/accounting/stats.go#L165). The transfer is running continuously and as you can see the elapsedTime and the transferTime already diverge massively after just around 90 minutes.

**Comment by JoshuaDietz** [comment]
Okay, here are more responses. All still within the 350K speed limit

After one more hour:
```
{
	"bytes": 3164372263,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 8829.01298152,
	"errors": 0,
	"fatalError": false,
	"renames": 0,
	"retryError": false,
	"speed": 2194403.5739601096,
	"transferTime": 1442.019569545,
	"transferring": [
		{
			"bytes": 3076096,
			"eta": 22,
			"group": "global_stats",
			"name": "backup1/my-devices/data/04/0428e1cb83f469cecb2500125a353159fe4382c326abfc29fd496f4bd05b46ae",
			"percentage": 59,
			"size": 5170515,
			"speed": 92717.95285677914,
			"speedAvg": 92062.50302345125
		},
		{
			"bytes": 2125824,
			"eta": 27,
			"group": "global_stats",
			"name": "backup1/my-devices/data/04/042a1016463e843041d6e30efcf0e3d1be6c4460ba0522b73ce04d058d579383",
			"percentage": 46,
			"size": 4607253,
			"speed": 91304.71442680797,
			"speedAvg": 90260.3594069796
		},
		{
			"bytes": 1634304,
			"eta": 29,
			"group": "global_stats",
			"name": "backup1/my-devices/data/04/042be42437fbd7fc2abf829cf738d031d2ca95eaef8069eb07bd4c6c27643424",
			"percentage": 38,
			"size": 4249153,
			"speed": 90630.46320407267,
			"speedAvg": 89198.5227501393
		},
		{
			"bytes": 159744,
			"eta": 132,
			"group": "global_stats",
			"name": "backup1/my-devices/data/04/043377c8a37eb1ecd0145fe1a6b04adc010bce407fafa2ed047840fd8d80840d",
			"percentage": 2,
			"size": 6403528,
			"speed": 72553.15120914545,
			"speedAvg": 47102.65059864635
		}
	],
	"transfers": 637
}

```

And after one more hour again:

```
{
	"bytes": 4472534371,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 12479.016624525,
	"errors": 0,
	"fatalError": false,
	"renames": 0,
	"retryError": false,
	"speed": 3157016.496211426,
	"transferTime": 1416.696800275,
	"transferring": [
		{
			"bytes": 5599232,
			"eta": 10,
			"group": "global_stats",
			"name": "backup1/my-devices/data/05/05e1da177aa3596b3c3bba334763849268b76a15459a3475882bd02f48f3fc47",
			"percentage": 84,
			"size": 6604122,
			"speed": 91151.91279626644,
			"speedAvg": 93015.0390133904
		},
		{
			"bytes": 2781184,
			"eta": 26,
			"group": "global_stats",
			"name": "backup1/my-devices/data/05/05e3c3b5eda10659d639df8ad1a25757f67ef505c015fed2fc6991ce556dd5ce",
			"percentage": 53,
			"size": 5164865,
			"speed": 92433.19332179548,
			"speedAvg": 90559.95051200983
		},
		{
			"bytes": 1961984,
			"eta": 41,
			"group": "global_stats",
			"name": "backup1/my-devices/data/05/05e7efd57d909ad098197e0d9ffe67546fba39ee4d34e81277a8d2f12309ed2e",
			"percentage": 34,
			"size": 5714158,
			"speed": 90129.12689590729,
			"speedAvg": 89491.94125638726
		},
		{
			"bytes": 815104,
			"eta": 54,
			"group": "global_stats",
			"name": "backup1/my-devices/data/05/05e873970ed25d1fdb9c857a11ce015d1828b1ab7553a1edb855f0c05f06fa7c",
			"percentage": 16,
			"size": 5058696,
			"speed": 84747.01408854034,
			"speedAvg": 78233.52067712341
		}
	],
	"transfers": 897
}


```

**Comment by ncw** [comment]
This has diverged massively though!

This says you've been running for 87 minutes but only transferring for 24 minutes.

> 	"elapsedTime": 5223.129277676,
> 	"transferTime": 1451.382347904,

And transfer time got smaller the next time

>	"elapsedTime": 12479.016624525,
>	"transferTime": 1416.696800275,

Are you doing anything else other than reading stats from this while it is running?

Are the jobs being put in via the RC or from the command line?

There is something kooky going on here definitely!

**Comment by ncw** [comment]
I managed to replicate this myself.

It starts going wrong at 100 transfers which at least pointed me at the bit of code that was wrong.

I eventually managed to figure out the problem - can you give this a go please?

If this fixes the problem I'll put it into v1.53.1

[v1.54.0-beta.4755.1a9a5d704.fix-4569-rc-stats](https://beta.rclone.org/branch/fix-4569-rc-stats/v1.54.0-beta.4755.1a9a5d704.fix-4569-rc-stats/) on branch [fix-4569-rc-stats](https://github.com/rclone/rclone/tree/fix-4569-rc-stats) (uploaded in 15-30 mins)

**Comment by JoshuaDietz** [comment]
> Are you doing anything else other than reading stats from this while it is running?
> 
> Are the jobs being put in via the RC or from the command line?

My setup looks like this: I have rclone running on a raspberry pi. It mounts a directory (my cloudstorage) which is used as a directory of a samba server. This way I can use a samba server to store files into the cloud without having to wait for the upload to complete. It works really well as far as I can see. I added a few hundred gigabytes about a week ago. Since my second run from above nothing was added, rclone is just continuing to upload its cache. Also I didn't use the rc to change anything, I only use it via the webui to monitor the upload progress.



> This says you've been running for 87 minutes but only transferring for 24 minutes.
Yes. But in this timeframe I uploaded `"bytes": 1872030707,`, which if divided by the `"elapsedTime": 5223.129277676` is `358 420`. This adds up almost perfectly with the 350K speed limit which was in place at that time. So the transfer was continuos.

**Comment by JoshuaDietz** [comment]
Oh sorry @ncw , I wrote my message before I saw your comment.

Your link to the beta does not seem to work, I just get a 404. I'll check back if it works in a few minutes just in case the build is still running.

**Comment by ncw** [comment]
> Your link to the beta does not seem to work, I just get a 404. I'll check back if it works in a few minutes just in case the build is still running.

It looks like everything built now.

**Comment by JoshuaDietz** [comment]
Yes now the link works. Unfortunately I can't test it right now because (with my previous configuration still running) I just get "Certificate signed by unknown authority" erros. Since my instance was still running and nothing changed I guess jottacloud is currently having some issues. I'll test it as soon as this is resolved.

**Comment by JoshuaDietz** [comment]
@ncw I tested it now. It looks good. After a bit more than an hour it looked like this:
```
{
	"bytes": 2400998155,
	"checks": 0,
	"deletes": 0,
	"elapsedTime": 4263.05890072,
	"errors": 0,
	"fatalError": false,
	"renames": 0,
	"retryError": false,
	"speed": 564026.8208737072,
	"transferTime": 4256.886826426,
	"transferring": [
		{
			"bytes": 4452352,
			"eta": 0,
			"group": "global_stats",
			"name": "backup1/my-devices/data/c0/c0111f10a944538c4cc6775f36a30ab23e96b34e46518710a78f87ee9be9701a",
			"percentage": 97,
			"size": 4573290,
			"speed": 144210.46159351376,
			"speedAvg": 143554.46788934097
		},
		{
			"bytes": 3698688,
			"eta": 3,
			"group": "global_stats",
			"name": "backup1/my-devices/data/c0/c016435aaa2685026bab8d875d08838264b315fe9f2da1096ac1e127ad390da8",
			"percentage": 86,
			"size": 4260625,
			"speed": 143270.02747783327,
			"speedAvg": 140556.97662090472
		},
		{
			"bytes": 2846720,
			"eta": 12,
			"group": "global_stats",
			"name": "backup1/my-devices/data/c0/c0171803fdca2ae00db77cfa266c5281df9e9ab6db92b13d3e0635b2f388f5f9",
			"percentage": 62,
			"size": 4561997,
			"speed": 141270.54188888904,
			"speedAvg": 139250.10237615206
		},
		{
			"bytes": 1044480,
			"eta": 31,
			"group": "global_stats",
			"name": "backup1/my-devices/data/c0/c0172b6b99397e611e96702fdb16ce514ecac0cc11fb8118256f61a1279ba275",
			"percentage": 20,
			"size": 5162507,
			"speed": 138474.00739411506,
			"speedAvg": 130485.9248728431
		}
	],
	"transfers": 523
}

```
I think this discrepancy is realistic especially if compared to the ones from above. So in my opinion it can be put into v1.53.1.

**Comment by ncw** [comment]
Thanks for testing :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

I'll stick it in v1.53.2 when that is released. You can try the stable branch here

[v1.53.2-beta.4766.b81dc1648.v1.53-stable](https://beta.rclone.org/branch/v1.53-stable/v1.53.2-beta.4766.b81dc1648.v1.53-stable/) on branch [v1.53-stable](https://github.com/rclone/rclone/tree/v1.53-stable) (uploaded in 15-30 mins)
